### PR TITLE
Added new class MainframeShell for mainframe(z/os)

### PR DESCRIPTION
**Pull request by bigendiansmalls**
This pull adds a mainframe shell class to the framework.  It also adds a spec to do a minimal test on the shell.

Both files in this pull are additions, no changes to existing files were made.   Essentially the mainframe shell overrides the basic CommandShell's methods responsible for reading and writing data to a target.  This is done so we can control the encoding/decoding of ascii<-> ebcdic data.

This pull has the following items:
- changes in lib/msf/base/sessions
  *`mainframe_shell.rb` - added support for MainframeShell class
- changes in spec/lib/msf/base/sessions
  *`mainframe_shell_spec.rb` - added spec for minimal testing of MainframeShell class
# RSpec / Validation
- [x] $ rspec -f doc spec/lib/msf/base/sessions/mainframe_shell_spec.rb

```
$ rspec spec/lib/msf/base/sessions/mainframe_shell_spec.rb 
Msf::Sessions::MainframeShell .

Finished in 0.03383 seconds
1 example, 0 failures

Randomized with seed 42315
```

Here is a screenshot of the implementation in a payload.
image

### CODE REVIEW

File: lib/msf/base/sessions/mainframe_shell.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  # override shell_read to include decode of cp1047
+  #
+  def shell_read(length=-1, timeout=1)
+    #mfimpl
+    if self.respond_to?(:ring)
+      return Rex::Text.from_ibm1047(shell_read_ring(length,timeout))
+    end
+
+    begin
+      rv = Rex::Ttext.from_ibm1047(rstream.get_once(length, timeout))
```
Comment (void-in) [inline lib/msf/base/sessions/mainframe_shell.rb]: A typo here. Should be Rex::Text.from_ibm1047(rstream.get_once(length, timeout))

Comment (bigendiansmalls) [inline lib/msf/base/sessions/mainframe_shell.rb]: Fixed.  Thank you!

**Comment by bigendiansmalls** [comment]
Hey all - not sure to whom to pose this; I've some more pulls waiting in the hopper that depend on this one (basic shell payloads etc).   Wondering if I should wait until this one lands before putting those up or just put them up and mention they're depending upon this pull?   Thanks!!

**Comment by wvu-r7** [comment]
@bigendiansmalls: We'll be on it!! Thank you for your patience and thanks for the poke! You can put up the other PRs and specify that they depend on this one. :)

**Comment by wvu-r7** [comment]
Grabbing this, since it's been sitting for so long. Sorry!

**Comment by bigendiansmalls** [comment]
Thanks a ton!  I didn't want  to be impatient :)

File: lib/msf/base/sessions/mainframe_shell.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    if self.respond_to?(:ring)
+      return Rex::Text.from_ibm1047(shell_read_ring(length,timeout))
+    end
+
+    begin
+      rv = Rex::Text.from_ibm1047(rstream.get_once(length, timeout))
+      framework.events.on_session_output(self, rv) if rv
+      return rv
+    rescue ::Rex::SocketError, ::EOFError, ::IOError, ::Errno::EPIPE => e
+      #print_error("Socket error: #{e.class}: #{e}")
```
Comment (wvu-r7) [inline lib/msf/base/sessions/mainframe_shell.rb]: Can you remove debugging lines like these?

Comment (wvu-r7) [inline lib/msf/base/sessions/mainframe_shell.rb]: Or I can just take care of it!

Comment (wvu-r7) [inline lib/msf/base/sessions/mainframe_shell.rb]: Handling it myself! Please don't commit. :)

Comment (bigendiansmalls) [inline lib/msf/base/sessions/mainframe_shell.rb]: LOL - no problem!  Thanks!

Comment (wvu-r7) [inline lib/msf/base/sessions/mainframe_shell.rb]: Okay, no problem, hah.

**Comment by wvu-r7** [comment]
@bigendiansmalls: You should use a topic branch instead of ```master``` next time, since I was going to PR changes to you.

**Comment by wvu-r7** [comment]
PR'ing anyway: https://github.com/bigendiansmalls/msf-mainframe-fork/pull/1.

**Comment by bigendiansmalls** [comment]
wvu-r7 - Ok - will do!  I'll check them out now and get back to you shortly.... thank you

**Comment by bigendiansmalls** [comment]
@wvu-r7 One question I posed on the PR in my fork, if you'd have a quick look when you are able.  Thanks.

**Comment by bigendiansmalls** [comment]
Should be good to go...

**Comment by wvu-r7** [comment]
Boom, it's in.

**Pull request merged**